### DISCUSSION TITLE
Update the context using values in an event

### DISCUSSION THREAD

**Discussion by daveyarwood** [statelyai/xstate#3468] [Q&A] [answered]
Hi, I'm new to XState and having a hard time figuring out the recommended way to update a context using values in an event.

My real world example is that I have a context which might have a Catalog (the definition of Catalog isn't important; it's an object with some properties):

```typescript
interface Context {
  catalog?: Catalog
}
```

I want to send a "select catalog" event that looks like this:

```typescript
{
  type: "SELECT_CATALOG",
  catalog: Catalog
}
```

This should result in the provided catalog from the event becoming the catalog in the context.

Based on the [Context page](https://xstate.js.org/docs/guides/context.html) in the docs, it looks like this is a job for assign, something like:

```typescript
const selectCatalog = assign({
  catalog: (_ctx, event) => event.catalog,
  message: "Selected catalog"
});
```

But that doesn't work because `Property 'catalog' does not exist on type 'EventObject'`. According to [the API docs](https://paka.dev/npm/xstate@4.32.1/api#ac096741838a4121), the _only_ property of EventObject is `type: string`, so it isn't clear to me how to access other properties of the event when going this route as recommended on the Context page.

The [Assign actions](https://xstate.js.org/docs/guides/context.html#assign-action) section has an example that references `event.value`, but that doesn't compile for me. I get `Property 'value' does not exist on type 'EventObject'.`, which makes sense based on the docs saying that `type: string` is the only property of EventObject. Is that example in the docs with `event.value` incorrect?

I noticed that if I instead supply a function to `assign`, it works, and the function operates on the actual event (i.e. my Event type, not EventObject). But then, it isn't clear to me how to update the context because I am not specifying (as above) the property of the context that I am updating. I read that mutating the context is to be avoided, so is the idea that my function is supposed to return an immutable copy of the context, like this?

```typescript
type Event =
  | { type: 'OPEN_FORM' }
  | { type: 'CLOSE_FORM' }
  | { type: 'CANCEL' }
  | { type: 'BACK' }
  | { type: 'NEXT_STEP' }
  | { type: 'SELECT_CATALOG', catalog: Catalog };

const selectCatalog = assign(
  (ctx : Context, event : Event) => {
    if (event.type !== "SELECT_CATALOG") return ctx;

    const ctxCopy = Object.assign({}, ctx);
    ctxCopy.catalog = event.catalog;
    return ctxCopy;
  }
);
```

Am I on the right track? Is there a better way to do this?

**Comment by christoph-fricke** [comment, marked as answer]
Hi there,

I am not sure why `assign` is not working for you, as it should work as expected. I created a little sandbox to show that your first example should indeed work. Updating context this way is a solid approach depending on the complexity of your context and update - at some point you **might** want to reach for _xstate-immer_.

Sandbox: https://codesandbox.io/s/delicate-darkness-8f9vpp

If it still does not work for you, can you share a reproducible example/sandbox of your problem? My guess is that the problem is not the `assign` action but rather the how/where the action is configured in the machine.

**Reply by daveyarwood** [reply]
Thanks, this is helpful!

This clearly works, but it's missing types, and when I try to add them in the sandbox, I get a similar error to the ones I was seeing with my code example.

For example, I changed this part:

```typescript
    actions: {
      assignCatalog: assign({ catalog: (ctx, e) => e.catalog })
    }
```

to:

```typescript
    actions: {
      assignCatalog: assign({ catalog: (ctx: Context, e: Event): Catalog => e.catalog })
    }
```

And the type checker complained that `Property 'catalog' does not exist on type 'Event'`.

**Reply by christoph-fricke** [reply]
@daveyarwood

**Notice the update at the bottom! I am keeping the original answer as well, since it contains useful blobs of information.**

The event param `e` should be typed with the event type for the machine and not the base event type, i.e. in the example sandbox `SelectCatalogEvent`. For multiple events, you should construct a union of all events that can be received by the machine.

```typescript
actions: {
    assignCatalog: assign({ catalog: (ctx: Context, e: SelectCatalogEvent): Catalog => e.catalog })
}
```

On a side note: The base type for events that are handled in machines is `EventObject` and not `Event`. `Event` can be the event type (used for `machine.send("SomeEvent")`) as well, but by the time the event is handled by your machine logic, it will always be resolved to an object (`EventObject`) by the interpreter.

However: In the sandbox, you **do not need to type the params at all**, since they are correctly inferred by TypeScript. The above (explicit types) only applies if you extract the action outside of the machine definition/configuration. Furthermore, you can use [`Typegen`](https://xstate.js.org/docs/guides/typescript.html#typegen) to let TypeScript infer event smarter types for your actions with the help of XState. It will constrain the event in actions and services to the exact events that actually trigger an action in the machine definition.

---

**UPDATE:** After reading the original question again, I am guessing that your `Event` is your custom type and not the provided `Event` type from XState. In that case, your comment is totally valid and you need another event check to constrain the event type in assign. I updated the [sandbox](https://codesandbox.io/s/delicate-darkness-8f9vpp?file=/src/index.ts) as well to reflect the change.

```typescript
actions: {
  assignCatalog: assign({
    catalog: (ctx: Context, e: Event) => // The explicit typing is optional
      e.type === "SELECT_CATALOG" ? e.catalog : ctx.catalog
  })
}
```

This is actually a perfect example of a problem that can be avoided with typegen. With typegen, the event type would be constrained to `SELECT_CATALOG` already and the ternary could be avoided.

**Reply by daveyarwood** [reply]
OK, I've figured it out, thanks in part to your code sandbox! :raised_hands: 

After making my code basically identical to yours, I was still getting a type check error, and then I realized that I was using `type Catalog = ...` instead of `interface Catalog { }`, and once I changed my code to use `interface` instead of `type`, the type errors went away. :thinking: 

I'm still fairly new to TypeScript, so now I guess to need to improve my understanding of when to use `type` vs. `interface`.

**Reply by daveyarwood** [reply]
Bizarrely, it only seems to work if my Catalog interface has only one thing, and the key must start with a lowercase letter, like in your example:

```typescript
interface Catalog {
  name: string;
}
```

The following cause my type error to return:

```typescript
interface Catalog {
  Name: string;
}
```

```typescript
interface Catalog {
  ImageKey: string;
  Name: string;
  SearchKeys: string[]
}
```

**Reply by daveyarwood** [reply]
Having a hard time repro-ing this in your sandbox though. Will continue to tinker with it.

**Reply by daveyarwood** [reply]
OK, here is a sandbox showing where I'm at now: https://codesandbox.io/s/beautiful-pare-lfxzge

There is a type error on line 47 where we attempt to include `selectCatalog` in the actions.

Strangely, I have the exact same code in my editor and I'm also seeing an _additional_ type error on line 23 that is not present in the sandbox:

```
app/components/add-products-machine.ts:23:12 - error TS2322: Type '(ctx: Context, e: Event) => Catalog | undefined' is not assignable to type 'Catalog | PartialAssigner<{ catalog: Catalog | undefined; message: string; }, EventObject, "catalog"> | undefined'.
  Type '(ctx: Context, e: Event) => Catalog | undefined' is not assignable to type 'PartialAssigner<{ catalog: Catalog | undefined; message: string; }, EventObject, "catalog">'.
    Types of parameters 'e' and 'event' are incompatible.
      Type 'EventObject' is not assignable to type 'Event'.
        Type 'EventObject' is not assignable to type '{ type: "ADD_TO_FLIGHT"; }'.
          Types of property 'type' are incompatible.
            Type 'string' is not assignable to type '"ADD_TO_FLIGHT"'.
```

I wonder why the type check results would be different in my project vs. in the sandbox? I checked, and they are both using the same version of XState, 4.32.1.

**Reply by daveyarwood** [reply]
Worth noting: if I copy-paste the `assign(...)` part of the `const selectProduct = assign(...)` directly into the `actions` part of the config, then it works. So my takeaway so far is that it is not possible to define these actions at the top level if I want to include types; they must be written inline as part of the `actions`.

EDIT: ...except I still can't get it working on my machine, even if it's working in the sandbox. This is so frustrating.

**Reply by daveyarwood** [reply]
Here is a sandbox that is more like my code: https://codesandbox.io/s/romantic-mcclintock-pur6tk?file=/src/index.ts

It works in the sandbox, but when I run `tsc` locally, I get this error:

```
app/components/add-products-machine.ts:103:9 - error TS2322: Type 'AssignAction<{ catalog: Catalog | undefined; message: string; }, Event>' is not assignable to type 'ActionObject<Context, Event> | ((context: Context, event: Event, meta: ActionMeta<Context, Event, BaseActionObject>) => void) | undefined'.
  Type 'AssignAction<{ catalog: Catalog | undefined; message: string; }, Event>' is not assignable to type 'ActionObject<Context, Event>'.
    Types of property 'exec' are incompatible.
      Type '((context: { catalog: Catalog | undefined; message: string; }, event: Event, meta: ActionMeta<{ catalog: Catalog | undefined; message: string; }, Event, BaseActionObject>) => void) | undefined' is not assignable to type '((context: Context, event: Event, meta: ActionMeta<Context, Event, BaseActionObject>) => void) | undefined'.
        Type '(context: { catalog: Catalog | undefined; message: string; }, event: Event, meta: ActionMeta<{ catalog: Catalog | undefined; message: string; }, Event, BaseActionObject>) => void' is not assignable to type '(context: Context, event: Event, meta: ActionMeta<Context, Event, BaseActionObject>) => void'.
          Types of parameters 'meta' and 'meta' are incompatible.
            Type 'ActionMeta<Context, Event, BaseActionObject>' is not assignable to type 'ActionMeta<{ catalog: Catalog | undefined; message: string; }, Event, BaseActionObject>'.
              The types of 'state.context' are incompatible between these types.
                Type 'Context' is not assignable to type '{ catalog: Catalog | undefined; message: string; }'.

103         selectCatalog: assign({
```

**Reply by daveyarwood** [reply]
My guess is that codesandbox.io is using a different version of TypeScript than I am. I am using TypeScript 4.2.2.

**Reply by daveyarwood** [reply]
Judging by [this](https://github.com/codesandbox/codesandbox-client/issues/5598), it looks like as of December 2021, codesandbox.io was using TypeScript 4.4.4. I tried using that version locally and got the same error above.

**Reply by daveyarwood** [reply]
I finally noticed in the lower right hand corner of codesandbox.io, it says "TypeScript 4.5.5." So I tried that locally, but I still got the same error above.

**Reply by daveyarwood** [reply]
I was able to repro the error I'm getting in [the XState visualizer](https://stately.ai/viz/3e4d3f40-b72e-4927-bb6d-595ec7e20e20). 

Note that it visualizes fine, but the error I posted above occurs on line 96.

**Reply by christoph-fricke** [reply]
@daveyarwood This is indeed one of the stranger problems, which may be a result of XState really pushing TypeScript's limits. Nonetheless, I encountered the problem before and fixed it. Left some general comment on some of your comments/code as well. ;)

```typescript
import { createMachine, State, assign, StateFrom } from "xstate";

interface Catalog {
  ImageKey?: string;
  NameKey: string;
  SearchKeys: string[];
}

interface Context {
  catalog?: Catalog;
  message?: string; // Required, if you want to assign it!
}

type Event =
  | { type: "OPEN_FORM" }
  | { type: "CLOSE_FORM" }
  | { type: "CANCEL" }
  | { type: "BACK" }
  | { type: "NEXT_STEP" }
  | { type: "SELECT_CATALOG"; catalog: Catalog }
  | { type: "ADD_TO_FLIGHT" };

// As far as I know, this does not provide much benefit, especially
// with Typegen available, and is no longer recommended.
type Typestate = {
  value: "form_closed" | "catalog_selection" | "products_selection" | "review";
  context: Context;
};

// Generic params are not needed when a schema is provided and
// providing a schema should be the preferred approach going forward.
const addProductsMachine = createMachine(
  // machine config
  {
    id: "addProductsMachine",
    // <original comment here>
    // This can be achieved with Typegen ;)
    initial: "form_closed",
    schema: {
      context: {} as Context,
      events: {} as Event,
    },
    context: {},
    // <original comment here>
    strict: true,
    states: {
      form_closed: {
        // <original comment here>
        // This can be achieved with Typegen ;)
        on: { OPEN_FORM: "form_open" },
      },
      form_open: {
        on: {
          CLOSE_FORM: "form_closed",
          CANCEL: "form_closed",
          ADD_TO_FLIGHT: {
            target: "form_closed",
            in: "form_open.review",
          },
        },
        initial: "catalog_selection",
        states: {
          catalog_selection: {
            on: {
              SELECT_CATALOG: {
                actions: ["selectCatalog"],
              },
              NEXT_STEP: "products_selection",
            },
          },
          products_selection: {
            on: {
              BACK: "catalog_selection",
              NEXT_STEP: "review",
            },
          },
          review: {
            on: {
              BACK: "products_selection",
            },
          },
        },
      },
    },
  },
  // machine options
  {
    actions: {
      selectCatalog: assign({
        catalog: (ctx, e) => {
          return e.type === "SELECT_CATALOG" ? e.catalog : ctx.catalog;
        },
        // XState is really pushing TS to is limits. For some reason, literal assignments sometimes do not work
        // in partial assignments. Therefore, this must be a assigner function, which strangely must accept two params.
        message: (_ctx, _e) => "selected catalog",
      }),
    },
  }
);

// <original comment here>
// type AddProductsFormState = State<Context, Event, any, Typestate, any>;

// This is a better and more robust approach for typing the whole machine state.
// In general, using *From (StateFrom, ActorFrom, ...) should be preferred over manually types.
type AddProductsFormState = StateFrom<typeof addProductsMachine>;

export { addProductsMachine, AddProductsFormState };
```

**Reply by daveyarwood** [reply]
Yes! I've finally got it working, thanks to your help. Thank you so much! :raised_hands: 

Your comments here are really helpful, too. I made a bunch of adjustments and simplifications, including removing the generic parameters in my `createMachine` call, and getting rid of `Typestate`.

One thing that finally clicked for me here was that I was under the impression from the docs that `message` was a special key that you could use with `assign` to give descriptions to events, and the value is expected to be a string. I didn't realize that it's no different from any other value in the context that you might be trying to assign. (i.e., if you want to include a message in the context, then you must add `message` to your Context type)

Thanks again for all your help! I think I'm basically un-stuck. The last thing I'm wondering about is: is it possible to add types to the actions? I tried this:

```typescript
  {
    actions: {
      selectCatalog: assign({
        catalog: (ctx: Context, e: Event): Catalog | undefined => {
          return e.type === "SELECT_CATALOG" ? e.catalog : ctx.catalog;
        }
      }),
    },
  }
```

But that resulted in a type error:

```
app/components/add-products-machine.ts:190:7 - error TS2322: Type 'AssignAction<{ selectedCreativeTemplate: CreativeTemplate | undefined; }, { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | ... 8 more ... | { ...; }>' is not assignable to type 'ActionObject<Context, { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | { type: "CATALOG_LOAD_ERROR"; } | { type: "CATALOGS_LOADED"; } | ... 6 more ... | { ...; }> | ((context: Context, event: { ...; } | ... 12 more ... | { ...; }, meta: ActionMeta<...'.
  Type 'AssignAction<{ selectedCreativeTemplate: CreativeTemplate | undefined; }, { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | ... 8 more ... | { ...; }>' is not assignable to type 'ActionObject<Context, { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | { type: "CATALOG_LOAD_ERROR"; } | { type: "CATALOGS_LOADED"; } | ... 6 more ... | { ...; }>'.
    Types of property 'exec' are incompatible.
      Type '((context: { selectedCreativeTemplate: CreativeTemplate | undefined; }, event: { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { ...; } | ... 8 more ... | { ...; }, meta: ActionMeta<...>) => void) | undefined' is not assignable to type '((context: Context, event: { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | { type: "CATALOG_LOAD_ERROR"; } | { type: "CATALOGS_LOADED"; } | ... 6 more ... | { ...; }, meta: ActionMeta<...>) => void) | undefined'.
        Type '(context: { selectedCreativeTemplate: CreativeTemplate | undefined; }, event: { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | ... 8 more ... | { ...; }, meta: ActionMeta<...>) => void' is not assignable to type '(context: Context, event: { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | { type: "CATALOG_LOAD_ERROR"; } | { type: "CATALOGS_LOADED"; } | ... 6 more ... | { ...; }, meta: ActionMeta<...>) => void'.
          Types of parameters 'meta' and 'meta' are incompatible.
            Type 'ActionMeta<Context, { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | { type: "CATALOG_LOAD_ERROR"; } | { type: "CATALOGS_LOADED"; } | ... 6 more ... | { ...; }, BaseActionObject>' is not assignable to type 'ActionMeta<{ selectedCreativeTemplate: CreativeTemplate | undefined; }, { type: "OPEN_FORM"; } | { type: "CLOSE_FORM"; } | { type: "CANCEL"; } | { type: "BACK"; } | { type: "NEXT_STEP"; } | ... 8 more ... | { ...; }, BaseActionObject>'.
              The types of 'state.context' are incompatible between these types.
                Type 'Context' is not assignable to type '{ selectedCreativeTemplate: CreativeTemplate | undefined; }'.
```

Here is the error [in the XState visualizer](https://stately.ai/viz/2ece3b4f-3214-4909-bf21-b0287aa17888). I copy-pasted your code above and attempted to add types to the `actions` part.

**Reply by christoph-fricke** [reply]
No problem. I am happy to help. 

> The last thing I'm wondering about is: is it possible to add types to the actions?

Yes. But the solution is weird... I think it falls into a similar category of weird TS behavior as the "message cannot be assigned as a literal" problem above.

It actually stops complaining when the third argument `meta`, which is provided to assigner functions, is included in the function signature as a parameter. However, it starts complaining again when `meta` is explicitly typed as `AssignMeta<Context, Event>` and I don't know why...

```typescript
{
    actions: {
      selectCatalog: assign({
        catalog: (ctx: Context, e: Event, meta): Catalog | undefined => {
          return e.type === "SELECT_CATALOG" ? e.catalog : ctx.catalog;
        }
      }),
    },
  }
```

I would recommend to not explicitly type the parameters and return type of assigner functions, since their types are already inferred perfectly by TS and it avoids these problems. Do you have specific reasons for explicitly typing them?

**Reply by daveyarwood** [reply]
Thanks, that seems like good advice.

I'm fairly new to TypeScript, and so far, I've been assuming that more explicit types == more type safety, but maybe I'm wrong about that. I still have a lot to learn!

**Reply by christoph-fricke** [reply]
In many cases explicit types are useful and lead to more type safety. At least for the cases where we create new types and function signatures. For example, when defining a function, types for the parameters are required but the return type is optional and can be inferred from the function body. However, defining the return type explicitly has the benefit of a clearer signature and enforces that all code paths in the function return the expected value.

In this case, the type of the expected callback is defined by `assign` and TypeScript can assert that any passed function must match the expected callback type. Anything else would be an error. Therefore, it already knows that the full function signature.
